cpp/src/arrow/csv/reader.cc

```diff
@@ -970,6 +972,145 @@ class StreamingReaderImpl : public ReaderMixin,
   std::shared_ptr<std::atomic<int64_t>> bytes_decoded_;
 };
 
+class ParallelStreamingReaderImpl
```

Comment:
I'm not found of all the duplication here. How about having `StreamingReaderImpl` take a buffer-async-generator instead?

Reply:
This code duplication is admittedly not ideal. However it's not just about reading faster through a new RandomAccessFile generator, but also parsing in parallel, and (in the future) decoding in parallel, which will introduce more changes to the StreamingReaderImpl. 

After I finish the decoder in parallel in a new PR perhaps then it makes sense to reduce the deduplication. Or perhaps I can just add it to this PR, though I think this PR has enough value to be merged by itself.

Reply:
This has been addressed.